File: repo-doctor-ts/src/analyzers/file-count.ts
import { BaseAnalyzer } from "./base";
import { readdir, stat } from "fs/promises";
import { join, extname } from "path";
import type { Finding } from "../types";

export class FileCountAnalyzer extends BaseAnalyzer {
  name = "file-count";
  description = "Counts files by type and flags repos with an unusually high file count";

  async run(repoPath: string) {
    const counts = new Map<string, number>();
    let totalFiles = 0;

    const walk = async (dir: string): Promise<void> => {
      const entries = await readdir(dir, { withFileTypes: true });
      for (const entry of entries) {
        if (entry.name.startsWith(".") || entry.name === "node_modules") continue;
        const fullPath = join(dir, entry.name);
        if (entry.isDirectory()) {
          await walk(fullPath);
        } else {
          totalFiles++;
          const ext = extname(entry.name) || "(no extension)";
          counts.set(ext, (counts.get(ext) || 0) + 1);
        }
      }
    };

    await walk(repoPath);

    const findings: Finding[] = [...counts.entries()]
      .sort((a, b) => b[1] - a[1])
      .map(([ext, count]) =>
        this.finding(repoPath, `${ext}: ${count} files`, "info")
      );

    const score = Math.max(0, Math.round(100 - (totalFiles / 5000) * 100));
    return this.result(findings, score);
  }
}

File: repo-doctor-ruby/lib/analyzers/file_count.rb
require_relative "base"

class FileCountAnalyzer < BaseAnalyzer
  def name = "file-count"
  def description = "Counts files by type and flags repos with high file count"

  def run(repo_path)
    counts = Hash.new(0)
    total = 0

    walk(repo_path) do |path|
      total += 1
      ext = File.extname(path).empty? ? "(no extension)" : File.extname(path)
      counts[ext] += 1
    end

    findings = counts.sort_by { |_, v| -v }.map do |ext, count|
      finding(file: repo_path, message: "#{ext}: #{count} files", severity: :info)
    end

    score = [0, (100 - (total.to_f / 5000 * 100)).round].max
    result(findings: findings, score: score)
  end

  private

  def walk(dir, &block)
    Dir.children(dir).each do |entry|
      next if entry.start_with?(".") || entry == "node_modules" || entry == "vendor"
      full_path = File.join(dir, entry)
      if File.directory?(full_path)
        walk(full_path, &block)
      else
        yield full_path
      end
    end
  end
end
